README.md

<!-- badges: start -->

[![CRAN\_Status\_Badge](https://www.r-pkg.org/badges/version-ago/aweek)](https://cran.r-project.org/package=aweek)
[![Codecov test
coverage](https://codecov.io/gh/reconhub/aweek/branch/master/graph/badge.svg)](https://codecov.io/gh/reconhub/aweek?branch=master)
[![Downloads](https://cranlogs.r-pkg.org/badges/grand-total/aweek?color=ff69b4)](https://cran.r-project.org/package=aweek)
[![DOI](https://zenodo.org/badge/172648833.svg)](https://zenodo.org/badge/latestdoi/172648833)
[![R-CMD-check](https://github.com/zkamvar/aweek/workflows/R-CMD-check/badge.svg)](https://github.com/zkamvar/aweek/actions)
<!-- badges: end -->

# Welcome to the *aweek* package\!

This package will convert dates to [US CDC
epiweeks](https://wwwn.cdc.gov/nndss/document/MMWR_Week_overview.pdf),
[isoweeks](https://en.wikipedia.org/wiki/ISO_week_date), and all others
in between with minimal overhead.

## Installing the package

To install the stable package version from CRAN, you can use

``` r
install.packages("aweek")
```

To benefit from the latest features and bug fixes, install the
development, *github* version of the package using:

``` r
# install.packages("remotes")
remotes::install_github("reconhub/aweek")
```

# Main Features

  - `date2week()` converts dates to a week format (YYYY-Www-d) that can
    start on any day.
  - `week2date()` / `as.Date()` does the backwards conversion from
    (YYYY-Www(-d)) to a numeric date.
  - `get_aweek()` generates an aweek object from a week number
  - `get_date()` converts a week number to a date
  - `as.aweek()` converts dates, characters, and factors to aweek
    objects.
  - `factor_aweek()` creates an aggregated factor of weeks where the
    levels contain all weeks within range.
  - Dependencies only on R itself.

<!-- -->

## Dates to weeks

With the *aweek* package, converting dates to weeks is simple. All you
need to know is what weekday represents the beginning of your week and a
vector of dates.

``` r
# generate dates
set.seed(2019-02-26)
onset <- as.Date("2019-02-26") + sort(sample(-6:7, 20, replace = TRUE))

# load aweek
library("aweek")
set_week_start("Monday") # set the default start of the week

print(onset)
```

    ##  [1] "2019-02-21" "2019-02-21" "2019-02-22" "2019-02-22" "2019-02-22" "2019-02-23" "2019-02-23" "2019-02-23" "2019-02-25" "2019-02-26" "2019-02-26" "2019-02-27" "2019-02-28" "2019-02-28" "2019-03-02" "2019-03-03"
    ## [17] "2019-03-04" "2019-03-05" "2019-03-05" "2019-03-05"

``` r
date2week(onset) # convert dates to weeks
```

    ## <aweek start: Monday>
    ##  [1] "2019-W08-4" "2019-W08-4" "2019-W08-5" "2019-W08-5" "2019-W08-5" "2019-W08-6" "2019-W08-6" "2019-W08-6" "2019-W09-1" "2019-W09-2" "2019-W09-2" "2019-W09-3" "2019-W09-4" "2019-W09-4" "2019-W09-6" "2019-W09-7"
    ## [17] "2019-W10-1" "2019-W10-2" "2019-W10-2" "2019-W10-2"

If you want to override the default, you can use the `week_start`
attribute of `date2week()`:

``` r
date2week(onset, week_start = 1) # ISO weeks starting on Monday (default)
```

    ## <aweek start: Monday>
    ##  [1] "2019-W08-4" "2019-W08-4" "2019-W08-5" "2019-W08-5" "2019-W08-5" "2019-W08-6" "2019-W08-6" "2019-W08-6" "2019-W09-1" "2019-W09-2" "2019-W09-2" "2019-W09-3" "2019-W09-4" "2019-W09-4" "2019-W09-6" "2019-W09-7"
    ## [17] "2019-W10-1" "2019-W10-2" "2019-W10-2" "2019-W10-2"

``` r
date2week(onset, week_start = 7) # EPI week starting on Sunday
```

    ## <aweek start: Sunday>
    ##  [1] "2019-W08-5" "2019-W08-5" "2019-W08-6" "2019-W08-6" "2019-W08-6" "2019-W08-7" "2019-W08-7" "2019-W08-7" "2019-W09-2" "2019-W09-3" "2019-W09-3" "2019-W09-4" "2019-W09-5" "2019-W09-5" "2019-W09-7" "2019-W10-1"
    ## [17] "2019-W10-2" "2019-W10-3" "2019-W10-3" "2019-W10-3"

``` r
date2week(onset, week_start = 6) # EPI week starting on Saturday
```

    ## <aweek start: Saturday>
    ##  [1] "2019-W08-6" "2019-W08-6" "2019-W08-7" "2019-W08-7" "2019-W08-7" "2019-W09-1" "2019-W09-1" "2019-W09-1" "2019-W09-3" "2019-W09-4" "2019-W09-4" "2019-W09-5" "2019-W09-6" "2019-W09-6" "2019-W10-1" "2019-W10-2"
    ## [17] "2019-W10-3" "2019-W10-4" "2019-W10-4" "2019-W10-4"

``` r
x <- date2week(onset, week_start = "sat", floor_day = TRUE) # truncate to just the weeks
table(x)
```

    ## x
    ## 2019-W08 2019-W09 2019-W10 
    ##        5        9        6

``` r
table(week2date(x))
```

    ## 
    ## 2019-02-16 2019-02-23 2019-03-02 
    ##          5          9          6

## Weeks to dates

If you have numeric weeks, you can rapidly convert to dates with
`get_date()`. Here are all the dates for the first day of last 10 ISO
weeks of 2015:

``` r
get_date(week = 44:53, year = 2015)
```

    ##  [1] "2015-10-26" "2015-11-02" "2015-11-09" "2015-11-16" "2015-11-23" "2015-11-30" "2015-12-07" "2015-12-14" "2015-12-21" "2015-12-28"

``` r
# you can also use this to generate aweek objects
get_aweek(week = 44:53, year = 2015)
```

    ## <aweek start: Monday>
    ##  [1] "2015-W44-1" "2015-W45-1" "2015-W46-1" "2015-W47-1" "2015-W48-1" "2015-W49-1" "2015-W50-1" "2015-W51-1" "2015-W52-1" "2015-W53-1"

If you have weeks recorded from different data sets that start on
different days, you can account for that by using the `start` option.
For example, 2018-01-01 is a Monday, but 2018-W01-1 starting on a Sunday
is 2017-12-31

``` r
get_date(week = 1, year = 2018, day = 1, start = c("Sunday", "Monday"))
```

    ## [1] "2017-12-31" "2018-01-01"

``` r
# get_aweek will align them to the default week_start
get_aweek(week = 1, year = 2018, day = 1, start = c("Sunday", "Monday"))
```

    ## <aweek start: Monday>
    ## [1] "2017-W52-7" "2018-W01-1"

## Factors

You can also automatically calculate factor levels, which is useful in
tabulating across weeks and including missing values.

``` r
set.seed(2019-02-26)
onset <- as.Date("2019-02-26") + sort(sample(-48:49, 20, replace = TRUE))
x     <- date2week(onset, week_start = "sat", factor = TRUE)
x
```

    ## <aweek start: Saturday>
    ##  [1] 2019-W03 2019-W03 2019-W04 2019-W04 2019-W04 2019-W05 2019-W05 2019-W05 2019-W06 2019-W06 2019-W06 2019-W09 2019-W09 2019-W10 2019-W11 2019-W12 2019-W12 2019-W14 2019-W15 2019-W15
    ## Levels: 2019-W03 2019-W04 2019-W05 2019-W06 2019-W07 2019-W08 2019-W09 2019-W10 2019-W11 2019-W12 2019-W13 2019-W14 2019-W15

``` r
table(x)
```

    ## x
    ## 2019-W03 2019-W04 2019-W05 2019-W06 2019-W07 2019-W08 2019-W09 2019-W10 2019-W11 2019-W12 2019-W13 2019-W14 2019-W15 
    ##        2        3        3        3        0        0        2        1        1        2        0        1        2

## Locales

It’s also possible to specify the week\_start in terms of the current
locale, however it is important to be aware that code like this may not
be portable.

``` r
# workaround because of differing locale specifications
german <- if (grepl("darwin", R.version$os)) "de_DE.UTF-8" else "de_DE.utf8"
lct <- Sys.getlocale("LC_TIME")
res <- Sys.setlocale("LC_TIME", german)

date2week(as.Date("2019-02-26"), week_start = "Sonntag")
```

    ## <aweek start: Sonntag>
    ## [1] "2019-W09-3"

``` r
date2week(as.Date("2019-02-26"), week_start = "Sunday")
```

    ## <aweek start: Sonntag>
    ## [1] "2019-W09-3"

``` r
Sys.setlocale("LC_TIME", lct)
```

    ## [1] "en_US.UTF-8"

## Getting help online

Bug reports and feature requests should be posted on *github* using the
[*issue*](https://github.com/reconhub/aweek/issues) system. All other
questions should be posted on the **RECON forum**: <br>
<https://www.repidemicsconsortium.org/forum/>

Contributions are welcome via **pull requests**.

Please note that this project is released with a [Contributor Code of
Conduct](CONDUCT.md). By participating in this project you agree to
abide by its terms.

## Similar Work

There are other packages that can define ISOweeks and/or epiweeks.
However, the ability to easily switch between day and week intervals is
only available for the ISOweek package and all of the packages above
have dependencies that require compiled code.

  - [ISOweek](https://cran.r-project.org/package=ISOweek) converts dates
    to ISO weeks as the “%W” and “%u” formats don’t exist in windows
  - [lubridate](https://github.com/tidyverse/lubridate) performs general
    datetime handling with some auxiliary functions that return the week
    or day of the week.
  - [surveillance](https://surveillance.r-forge.r-project.org/)
    implements ISOWeekYear.


NEWS.md
# aweek 1.0.3

## CRAN MAINTENANCE

* `as.POSIXlt()` method for aweek objects no longer converts dates to character
  before converting to POSIXlt

# aweek 1.0.2

## CRAN MAINTENANCE

* A test that was failing due to an incorrect assumption that the first day of
  the year would always fall on the first week of the year was fixed 
  (see https://github.com/reconhub/aweek/issues/35 for details).
  (#36, @zkamvar).

## MISC

* Continuous integration has been migrated from Travis-CI + AppVeyor to
  GitHub actions.

# aweek 1.0.1

## CRAN MAINTENANCE

* `as.POSIXlt()` now converts NULL to an empty POSIXlt object, so a test that
  was previously checking for an error failed. That test has been fixed
  (#33, @zkamvar)

# aweek 1.0.0

## NEW FUNCTIONS

* `change_week_start()` allows the user to change the `week_start` attribute of
  an aweek object, adjusting the weeks to match the new attribute accordingly.
* `get_aweek()` can generate aweek objects from a vector of week numbers. It has
  the ability to take into account different week start times. 
* `get_date()` is similar to `get_aweek()`, but returns dates instead.
* `as.aweek()` allows users to create aweek object directly from characters with
  validation. It also allows for dates by passing to `date2week()`.
* `as.data.frame.aweek()` is a new function that allows aweek objects to be
  directly incorporated into data frames.
* `as.list.aweek()` will now preserve the aweek structure in lists
* `trunc.aweek()` will truncate the day to the first day of the week. 
* `rep.aweek()` allows repeating aweek characters.
* `factor_aweek()` allows the user to create aggregated aweek objects on the fly.

## BREAKING CHANGES

There are a couple of breaking changes coming to aweek that will improve
stability by removing unclear coercion methods
(see https://github.com/reconhub/aweek/issues/20).

* It is no longer possible to combine aweek objects if they do not share the
  same `week_start` attribute. This will result in an error informing the user
  to adjust the `week_start` attribute with the `change_week_start()` function.
* Factors will no longer coerce factors when combining aweek objects. 
* Using `date2week()` with `factor = TRUE` and `floor_day = FALSE` now throws an
  error instead of a warning (as prophesized in #13).

## DOCUMENTATION

* The vignette has been updated to include an explanation of the underlying
  week calculation from date.
* The aweek class documentation has been updated to detail the different
  elements of the object and the calculations. 
* Package documentation `package?aweek` has been added for an introduction.

## BUG FIX

* conversions will now retain the names of the object.

## MISC

* More checks that weekdays and weeks are within the correct bounds have been
  added. 
* as.Date.aweek has been simplified to no longer rely on regex since the aweek
  class is standard.
* The internal `get_wday()` has been vastly simplified with improved speed.

# aweek 0.3.0

* The `week_start` argument now defaults to the global option `aweek.week_start`,
  which will be a number from 1 to 7, representing the days of the week in the
  ISO 8601 standard.
* `set_week_start()` is a convenience allowing the user to set the default 
  `aweek.week_start` option via integer or character input.
* `get_week_start()` is a wrapper for `getOption("aweek.week_start", 1L)` and
  `attr(w, "week_start")` for aweek objects.

# aweek 0.2.2

* Simplified conversion to factors.
* Using `factor = TRUE` without `floor_day = TRUE` will now issue a message
  indicating that this is deprecated in future versions of aweek (see #13).

# aweek 0.2.1

* Fix bug where NAs threw errors in the dates (found: @aspina7, #12)

* `as.data.frame.aweek()` will now convert aweek objects to columns of data
  frames without losing class or attributes
* The introduction vignette has been updated to reflect this change. 

# aweek 0.2.0

* Subsetting and concatenating methods added to the `aweek` class (see #1)
* Documentation divided into smaller chunks
* `as.POSIXlt()` bug where `tz` was not being passed was fixed.
* `date2week()`: an error is now issued if users specify non-ISO 8601 dates OR
  don't specify a `format` option. (found: @scottyaz, #2)
* Best practices added to vignette
* Fix test that would fail every seven days on CRAN

# aweek 0.1.0

* First official release of aweek on CRAN

# aweek 0.0.5.9000

* `date2week()` and `week2date()` can now take days represented as characters in
  the current or English locale. 

# aweek 0.0.4.9000

* `date2week()` gains a `factor` argument, which will automatically compute the
  levels within the date range.

# aweek 0.0.3.9000

* `date2week()` now properly accounts for dates in December that occur in the
  first week of the next year. 

# aweek 0.0.2.9000

* `floor_day` now truncates the week instead changing the last digit to 1 for
  aesthetics. (Thanks to @aspina7 for the suggestion)
q `print.aweek()` now displays the day of the week in the current locale.

# aweek 0.0.1.9000

* First version of package
* `date2week()` converts dates to `aweek` objects
* `week2date()` converts `aweek` objects or character strings to dates
* `as.Date()` does the same thing as above
* `as.POSIXlt()` as well
* `as.character()` will unclass the `aweek` object
* Added a `NEWS.md` file to track changes to the package.


R/date2week.R
#' Convert date to a an arbitrary week definition
#'
#' @param x a [Date], [POSIXt], [character], or any data that can be easily
#'   converted to a date with [as.POSIXlt()]. 
#'
#' @param week_start a number indicating the start of the week based on the ISO
#'   8601 standard from 1 to 7 where 1 = Monday OR an abbreviation of the
#'   weekdate in an English or current locale. _Note: using a non-English locale
#'   may render your code non-portable._ Defaults to the value of 
#'   [get_week_start()]
#' 
#' @param floor_day when `TRUE`, the days will be set to the start of the week.
#'
#' @param numeric if `TRUE`, only the numeric week be returned. If `FALSE`
#'   (default), the date in the format "YYYY-Www-d" will be returned.  
#'
#' @param factor if `TRUE`, a factor will be returned with levels spanning the
#'   range of dates. This should only be used with `floor_day = TRUE` to
#'   produce the sequence of weeks between the first and last date as the
#'   factor levels.  Currently, `floor_date = FALSE` will still work, but will
#'   produce a message indicating that it is deprecated. _Take caution when
#'   using this with a large date range as the resulting factor can contain all
#'   days between dates_.
#' 
#' @param ... arguments passed to [as.POSIXlt()], unused in all other cases.
#' 
#' @details Weeks differ in their start dates depending on context. The ISO
#'   8601 standard specifies that Monday starts the week
#'   (<https://en.wikipedia.org/wiki/ISO_week_date>) while the US CDC uses
#'   Sunday as the start of the week
#'   (<https://stacks.cdc.gov/view/cdc/22305>). For
#'   example, MSF has varying start dates depending on country in order to
#'   better coordinate response. 
#'
#'   While there are packages that provide conversion for ISOweeks and epiweeks,
#'   these do not provide seamless conversion from dates to epiweeks with 
#'   non-standard start dates. This package provides a lightweight utility to
#'   be able to convert each day.
#'
#' @return
#'  - `date2week()` an [aweek][aweek-class] object which represents dates in
#'    `YYYY-Www-d` format where `YYYY` is the year (associated with the week,
#'    not necessarily the day), `Www` is the week number prepended by a "W" that
#'    ranges from 01-53 and `d` is the day of the week from 1 to 7 where 1
#'    represents the first day of the week (as defined by the `week_start`
#'    attribute).
#'  - `week2date()` a [Date][Date] object.  
#' 
#' @note `date2week()` will initially convert the input with [as.POSIXlt()] and
#'   use that to calculate the week. If the user supplies character input, it
#'   is expected that the input will be of the format yyyy-mm-dd _unless_ the 
#'   user explicitly passes the "format" parameter to [as.POSIXlt()]. If the
#'   input is not in yyyy-mm-dd and the format parameter is not passed, 
#'   `date2week()` will result in an error.
#'
#' @author Zhian N. Kamvar
#' @export
#' @seealso [set_week_start()], [as.Date.aweek()], [print.aweek()], [as.aweek()],
#'   [get_aweek()]
#' @examples
#'
#' ## Dates to weeks -----------------------------------------------------------
#'
#' # The same set of days will occur in different weeks depending on the start
#' # date. Here we can define a week before and after today
#'
#' print(dat <- as.Date("2018-12-31") + -6:7)
#' 
#' # By default, the weeks are defined as ISO weeks, which start on Monday
#' print(iso_dat <- date2week(dat))
#' 
#' # This can be changed by setting the global default with set_week_start()
#' 
#' set_week_start("Sunday")
#'
#' date2week(dat)
#'
#' # If you want lubridate-style numeric-only weeks, you need look no further
#' # than the "numeric" argument
#' date2week(dat, numeric = TRUE)
#' 
#' # To aggregate weeks, you can use `floor_day = TRUE`
#' date2week(dat, floor_day = TRUE)
#'
#' # If you want aggregations into factors that include missing weeks, use
#' # `floor_day = TRUE, factor = TRUE`:
#' date2week(dat[c(1, 14)], floor_day = TRUE, factor = TRUE)
#'
#'
#' ## Weeks to dates -----------------------------------------------------------
#'
#' # The aweek class can be converted back to a date with `as.Date()`
#' as.Date(iso_dat)
#' 
#' # If you don't have an aweek class, you can use week2date(). Note that the
#' # week_start variable is set by the "aweek.week_start" option, which we will
#' # set to Monday:
#' 
#' set_week_start("Monday")
#' week2date("2019-W01-1") # 2018-12-31
#'
#' # This can be overidden by the week_start argument;
#' week2date("2019-W01-1", week_start = "Sunday") # 2018-12-30
#'
#' # If you want to convert to the first day of the week, you can use the 
#' # `floor_day` argument
#' as.Date(iso_dat, floor_day = TRUE)
#'
#' ## The same two week timespan starting on different days --------------------
#' # ISO week definition: Monday -- 1
#' date2week(dat, 1)
#' date2week(dat, "Monday")
#'
#' # Tuesday -- 2
#' date2week(dat, 2)
#' date2week(dat, "Tuesday")
#'
#' # Wednesday -- 3
#' date2week(dat, 3)
#' date2week(dat, "W") # you can use valid abbreviations
#'
#' # Thursday -- 4
#' date2week(dat, 4)
#' date2week(dat, "Thursday")
#'
#' # Friday -- 5
#' date2week(dat, 5)
#' date2week(dat, "Friday")
#'
#' # Saturday -- 6
#' date2week(dat, 6)
#' date2week(dat, "Saturday")
#'
#' # Epiweek definition: Sunday -- 7 
#' date2week(dat, 7)
#' date2week(dat, "Sunday")
date2week <- function(x, week_start = get_week_start(), floor_day = factor, numeric = FALSE, factor = FALSE, ...) {

  # Cromulence checks ----------------------------------------------------------
  format_exists <- !is.null(list(...)$format)
  nas  <- is.na(x)
  nams <- names(x)

  week_start <- parse_week_start(week_start)

  if (!inherits(x, "aweek") && is.character(x) && !format_exists) {
    iso_std <- grepl("^[0-9]{4}[^[:alnum:]]+[01][0-9][^[:alnum:]]+[0-3][0-9]$", trimws(x))
    iso_std[nas] <- TRUE # prevent false alarms
    if (!all(iso_std)) {
      msg <- paste("Not all dates are in ISO 8601 standard format (yyyy-mm-dd).",
                   "The first incorrect date is %s"
      )
      stop(sprintf(msg, x[!iso_std][1]))
    }
  }

  x  <- tryCatch(as.POSIXlt(x, ...), error = function(e) e)

  if (inherits(x, "error")) {
    mc <- match.call()
    msg <- "%s could not be converted to a date. as.POSIXlt() returned this error:\n%s"
    stop(sprintf(msg, deparse(mc[["x"]]), x$message))
  }

  # Conversion from date to week -----------------------------------------------
  #
  # Step 1: Get the weekday as an integer in pseudo ISO 8601 ---------
  wday     <- as.integer(x$wday) # weekdays in R run 0:6, 0 being Sunday
  the_date <- as.Date.POSIXlt(x)

  # Step 2: Find week and weekday ------------------------------------ 
  # This part is important for accurately assessing the week. It shifts the date
  # to the middle of the relative week so that it can accurately be counted.
  # 
  # If the weekday is in the first three days of the week, then the middle of
  # the week is in the future, but if the weekday is in the last three days of
  # the week, then the middle of the week has already past and the date needs
  # to be shifted backward.
  wday     <- get_wday(wday, week_start)
  midweek  <- the_date + (4L - wday)
  the_week <- week_in_year(midweek)

  if (!numeric) {
    # Adding years to weeks ----------------------------------------------------
    # Step 1: find the months of each date -----------------------------
    december <- format(the_date, "%m") == "12"
    january  <- format(the_date, "%m") == "01"
    boundary_adjustment <- integer(length(the_date))
    
    # Step 2: Shift the years based on month/week agreement ------------
    # Shift the year backwards if the date is in january, but the week is not
    boundary_adjustment[january  & the_week >= 52] <- -1L
    # Shift the year forwards if the date is in december, but it's the first week
    boundary_adjustment[december & the_week == 1]  <- 1L

    # Step 3: Create ISO week strings ----------------------------------
    the_year <- as.integer(format(the_date, "%Y"))
    the_week <- sprintf("%04d-W%02d-%d", 
                        the_year + boundary_adjustment,
                        the_week,
                        wday
                       )
    # set the missing data back to missing -----------------------------
    the_week[nas] <- NA

    class(the_week) <- c("aweek", oldClass(the_week))
    attr(the_week, "week_start") <- week_start 

    if (floor_day) {
      the_week <- trunc(the_week)
    }

    if (factor) {
      if (!floor_day) {
        stop("as of aweek 1.0, using factor without floor_day is not allowed.")
      }
      the_week <- factor_aweek(the_week)
    }

  }
  names(the_week) <- names(x)
  the_week
}

